### PR TITLE
Add atscfg logic for Server Capabilities

### PR DESCRIPTION
**Pull request by rob05c**
Adds parent.config omitting parents which lack server capabilities required by Delivery Services.

- [x] This PR is not related to any other Issue

Includes changelog.
Includes unit tests for config generation.
Documentation already exists for Server Capabilities; ATS config generation details are not currently documented.

## Which Traffic Control components are affected by this PR?
- Traffic Ops
- Traffic Ops ORT

## What is the best way to verify this PR?
Run unit tests.

Create a DS with Required Capabilities, ensure servers exist which lack those Capabilities, verify parent.config (both at the old TO endpoint, and by `atstccfg`) don't contain those servers. Verify parent.config DSes with Required Capabilities do contain servers which have the Required Capabilities.

## If this is a bug fix, what versions of Traffic Control are affected?
Not a bug fix.

## The following criteria are ALL met by this PR

- [x] This PR includes tests OR I have explained why tests are unnecessary
- [x] This PR includes documentation OR I have explained why documentation is unnecessary
- [x] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR ensures that database migration sequence is correct OR this PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

File: lib/go-atscfg/parentdotconfig.go [hunk 1]
```diff
@@ -405,17 +338,119 @@ func MakeParentDotConfig(
...
 	profileCaches map[ProfileID]ProfileCache, // getServerParentCacheGroupProfiles(tx, server)
-	originServers map[string][]CGServer, // getServerParentCacheGroupProfiles(tx, server)
-) map[string][]ParentInfo {
-	parentInfos := map[string][]ParentInfo{}
+	originServers map[OriginHost][]CGServer, // getServerParentCacheGroupProfiles(tx, server)
+) map[OriginHost][]ParentInfo {
+	parentInfos := map[OriginHost][]ParentInfo{}
 
 	// note servers also contains an "all" key
 	// originFQDN is "prefix" in Perl; ds is not really a "ds", that's what it's named in Perl
```
Comment (awilliams) [inline lib/go-atscfg/parentdotconfig.go]: Does `originFQDN` in the comment need updating?

File: lib/go-atscfg/parentdotconfig.go [hunk 2]
```diff
@@ -455,3 +491,12 @@ func unavailableServerRetryResponsesValid(s string) bool {
 	re := regexp.MustCompile(`^"(:?\d{3},)+\d{3}"\s*$`) // TODO benchmark, cache if performance matters
 	return re.MatchString(s)
 }
+
```
Comment (awilliams) [inline lib/go-atscfg/parentdotconfig.go]: Consider providing a function comment.

File: lib/go-atscfg/parentdotconfig.go (hunk 1 above)
Comment (rob05c) [reply lib/go-atscfg/parentdotconfig.go]: I removed it. I think it's diverged from the old Perl enough to no longer apply.

File: lib/go-atscfg/parentdotconfig.go (hunk 2 above)
Comment (rob05c) [reply lib/go-atscfg/parentdotconfig.go]: Done.

File: traffic_ops/ort/atstccfg/toreq.go [hunk 3]
```diff
@@ -406,3 +407,74 @@ func GetJobs(cfg TCCfg) ([]tc.Job, error) {
...
+func GetServerCapabilitiesByID(cfg TCCfg, serverIDs []int) (map[int]map[atscfg.ServerCapability]struct{}, error) {
+	serverIDsStr := ""
+	if len(serverIDs) > 0 {
+		sortIDsInHash := true
+		serverIDsStr = base64.RawURLEncoding.EncodeToString((util.HashInts(serverIDs, sortIDsInHash)))
+	}
+
+	serverCaps := map[int]map[atscfg.ServerCapability]struct{}{}
+	err := GetCachedJSON(cfg, "server_capabilities_s_"+serverIDsStr+".json", &serverCaps, func(obj interface{}) error {
+		// TOO add list of IDs to API+Client
```
[the remark this answers is not in the document]
Comment (rob05c) [reply traffic_ops/ort/atstccfg/toreq.go]: Done.

File: traffic_ops/ort/atstccfg/toreq.go [hunk 4]
```diff
@@ -406,3 +407,74 @@ func GetJobs(cfg TCCfg) ([]tc.Job, error) {
...
+func GetDeliveryServiceRequiredCapabilitiesByID(cfg TCCfg, dsIDs []int) (map[int]map[atscfg.ServerCapability]struct{}, error) {
+	dsIDsStr := ""
+	if len(dsIDs) > 0 {
+		sortIDsInHash := true
+		dsIDsStr = base64.RawURLEncoding.EncodeToString((util.HashInts(dsIDs, sortIDsInHash)))
+	}
+
+	dsCaps := map[int]map[atscfg.ServerCapability]struct{}{}
+	err := GetCachedJSON(cfg, "ds_capabilities_d_"+dsIDsStr+".json", &dsCaps, func(obj interface{}) error {
+		// TOO add list of IDs to API+Client
```
[the remark this answers is not in the document]
Comment (rob05c) [reply traffic_ops/ort/atstccfg/toreq.go]: Done.

**Pull request merged**